math/WaveConversion.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2020 Nicholas Corgan
+// SPDX-License-Identifier: BSL-1.0
+
+#include <Pothos/Exception.hpp>
+#include <Pothos/Framework.hpp>
+
+#include <Poco/Format.h>
+
+#include <algorithm>
+#include <complex>
+#include <cmath>
+#include <string>
+
+template <typename T>
+static T waveTodB(const std::complex<T>& wave)
+{
+    return T(20.0) * std::log10(std::abs(wave));
+}
+
+template <typename T>
+static T waveTodBm(const std::complex<T>& wave)
+{
+    return T(10.0) * std::log10(std::abs(wave));
```

Comment:
dB is relative, dBm is absolute power, but the multiply by 20 or 10 has nothing to do with the unit nomenclature. 10 is for power (the square is kind of already in there) and 20 is for voltage.

You can also save a square root here if you feed the magnitude squared to the log10, and pull a 2 out of the first factor.

Reply:
My understanding is that these were the typical conversions to get from a complex value to these units. At least that's what I've seen and used everywhere. Could you clarify?

See: https://www.mathworks.com/help/signal/ref/mag2db.html

Reply:
10log10 is for converting power to decibels, and 20log10 is for converting voltage. The idea is that if you double your amplitude, you would see about a 3db increase. Well if you double your amplitude, you quadruple your power, so that whole 10log10 thing is really just compensating for a square. `P=V^2/R`

I haven't seen the 10log10 of a complex value, unless I guess someone pre-computed the magnitude squared. You know how those folks like to avoid square roots when possible.

In the case that you have power measured in milliwatts, then the units will be dBm.

I have also seen dBfs where its decibels full-scale. This is where you know that the maximum value of the ADC (say 1 << 15 for 16-bit integers or 1.0 when scaled with floats). So you do 20log10 and add in an adjustment for a specified reference level. The fft plotter block does this.

You could also calibrate this adjustment to have the units in dBm.